File: app/src/main/java/sr/otaryp/tesatyla/presentation/ui/custom/CircularProgressBar.java
package sr.otaryp.tesatyla.presentation.ui.custom;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RadialGradient;
import android.graphics.Shader;
import android.util.AttributeSet;
import android.view.View;

public class CircularProgressBar extends View {

    private Paint backgroundPaint;
    private Paint progressPaint;
    private float progress = 0f; // Start with 0% progress
    private int strokeWidth = 20; // Set stroke width
    private float radius;
    
    // Constructor
    public CircularProgressBar(Context context) {
        super(context);
        init();
    }

    public CircularProgressBar(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        // Paint for background circle (inactive portion)
        backgroundPaint = new Paint();
        backgroundPaint.setColor(Color.parseColor("#E0E0E0"));
        backgroundPaint.setStyle(Paint.Style.STROKE);
        backgroundPaint.setStrokeWidth(strokeWidth);
        backgroundPaint.setAntiAlias(true);

        // Paint for progress circle (active portion)
        progressPaint = new Paint();
        progressPaint.setStyle(Paint.Style.STROKE);
        progressPaint.setStrokeWidth(strokeWidth);
        progressPaint.setAntiAlias(true);

        // Apply gradient to progress paint
        RadialGradient gradient = new RadialGradient(0, 0, radius, Color.parseColor("#1A7F7F"), Color.parseColor("#009999"), Shader.TileMode.CLAMP);
        progressPaint.setShader(gradient);

        // Animate progress (example)
        ValueAnimator animator = ValueAnimator.ofFloat(0, 360); // Animate from 0% to 100%
        animator.setDuration(2000); // Duration of animation (2 seconds)
        animator.addUpdateListener(animation -> {
            progress = (float) animation.getAnimatedValue();
            invalidate(); // Redraw view to show progress update
        });
        animator.start();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        radius = Math.min(w, h) / 2f - strokeWidth; // Set radius based on view size
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        
        // Draw background circle
        canvas.drawCircle(getWidth() / 2f, getHeight() / 2f, radius, backgroundPaint);

        // Draw progress circle (animated)
        float angle = (360 * progress) / 100;
        canvas.drawArc(
                strokeWidth, strokeWidth, getWidth() - strokeWidth, getHeight() - strokeWidth, 
                -90, angle, false, progressPaint); // -90 to start the arc from the top
    }

    // Method to update progress dynamically
    public void setProgress(float progress) {
        this.progress = progress;
        invalidate(); // Redraw the view with updated progress
    }
}

File: app/src/main/java/sr/otaryp/tesatyla/presentation/ui/custom/CircularProgressBar.kt
package sr.otaryp.tesatyla.presentation.ui.custom

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.SweepGradient
import android.util.AttributeSet
import android.view.View
import android.view.animation.LinearInterpolator

class CircularProgressBar @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
) : View(context, attrs) {

    private val backgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val progressPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var progress: Float = 0f
    private val strokeWidth: Float = resources.displayMetrics.density * 16f
    private var radius: Float = 0f
    private val maxProgress = 100f

    private var currentStartAngle = -90f
    private var staticStartAngle: Float? = null
    private val spinSweepAngle = 270f
    private var isIndeterminate = false
    private val gradientMatrix = Matrix()
    private var progressShader: SweepGradient? = null

    private val rotationAnimator: ValueAnimator = ValueAnimator.ofFloat(0f, 360f).apply {
        duration = 1_500L
        interpolator = LinearInterpolator()
        repeatCount = ValueAnimator.INFINITE
        addUpdateListener { animator ->
            currentStartAngle = (animator.animatedValue as Float) - 90f
            staticStartAngle = currentStartAngle
            invalidate()
        }
    }

    init {
        backgroundPaint.apply {
            color = Color.TRANSPARENT
            style = Paint.Style.STROKE
            strokeWidth = this@CircularProgressBar.strokeWidth
        }

        progressPaint.apply {
            style = Paint.Style.STROKE
            strokeWidth = this@CircularProgressBar.strokeWidth
            strokeCap = Paint.Cap.ROUND
        }
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        radius = (minOf(w, h) / 2f) - (strokeWidth / 2f)

        progressShader = SweepGradient(
            w / 2f,
            h / 2f,
            intArrayOf(
                Color.parseColor("#1DF2F8"),
                Color.parseColor("#00F8FF"),
                Color.parseColor("#001DF2F8"),
            ),
            floatArrayOf(0f, 0.72f, 1f),
        ).also(progressPaint::setShader)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        if (backgroundPaint.color != Color.TRANSPARENT) {
            canvas.drawCircle(width / 2f, height / 2f, radius, backgroundPaint)
        }

        val startAngle = when {
            isIndeterminate -> currentStartAngle
            staticStartAngle != null -> staticStartAngle!!
            else -> -90f
        }

        val sweepAngle = when {
            isIndeterminate || staticStartAngle != null -> spinSweepAngle
            else -> (360f * progress) / maxProgress
        }

        progressShader?.let { shader ->
            gradientMatrix.reset()
            gradientMatrix.postRotate(startAngle + 90f, width / 2f, height / 2f)
            shader.setLocalMatrix(gradientMatrix)
        }

        canvas.drawArc(
            strokeWidth / 2f,
            strokeWidth / 2f,
            width - (strokeWidth / 2f),
            height - (strokeWidth / 2f),
            startAngle,
            sweepAngle,
            false,
            progressPaint,
        )
    }

    fun setProgress(progress: Float) {
        this.progress = progress.coerceIn(0f, maxProgress)
        staticStartAngle = null
        invalidate()
    }

    fun startIndeterminateAnimation() {
        if (isIndeterminate) return
        isIndeterminate = true
        rotationAnimator.start()
    }

    fun stopIndeterminateAnimation(resetToStart: Boolean = false) {
        if (!isIndeterminate && !rotationAnimator.isRunning) {
            if (resetToStart) {
                staticStartAngle = null
                currentStartAngle = -90f
                invalidate()
            }
            return
        }

        rotationAnimator.cancel()
        isIndeterminate = false

        if (resetToStart) {
            staticStartAngle = null
            currentStartAngle = -90f
        }

        invalidate()
    }

    override fun onDetachedFromWindow() {
        rotationAnimator.cancel()
        super.onDetachedFromWindow()
    }
}
